src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py

```diff
@@ -0,0 +1,114 @@
+"""
+_GetCompletedBlocks_
+
+MySQL implementation of RucioInjector.GetClompletedBlocks
+
+Retrieve a list of blocks that are complete,
+including their location and the sites they
+are subscribed too
+
+"""
+
+from __future__ import division
+from __future__ import print_function
+
+from WMCore.Database.DBFormatter import DBFormatter
+
+
+class GetCompletedBlocks(DBFormatter):
+    """
+    Retrieves a list of blocks that are closed but sure if deletedable:
+      - workflow completions for all files in the block to be checked on a later stage
+      - subscription made for dataset is copy+delete
+      - subscription has been made at the Data Management system
+      - blocks hasn't been deleted yet
+    """
+    sql = """SELECT dbsbuffer_block.blockname,
+                    dbsbuffer_location.pnn,
+                    dbsbuffer_dataset.path,
+                    dbsbuffer_dataset_subscription.site,
+                    dbsbuffer_workflow.name,
+                    dbsbuffer_block.create_time
+             FROM dbsbuffer_dataset_subscription
+             INNER JOIN dbsbuffer_dataset ON
+               dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
+             INNER JOIN dbsbuffer_block ON
+               dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
+             INNER JOIN dbsbuffer_file ON
+               dbsbuffer_file.block_id = dbsbuffer_block.id
+             INNER JOIN dbsbuffer_workflow ON
+               dbsbuffer_workflow.id = dbsbuffer_file.workflow
+             INNER JOIN dbsbuffer_location ON
+               dbsbuffer_location.id = dbsbuffer_block.location
+             WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
+             AND dbsbuffer_dataset_subscription.subscribed = 1
+             AND dbsbuffer_block.status = 'Closed'
+             AND dbsbuffer_block.deleted = 0
+             GROUP BY dbsbuffer_block.blockname,
```

Comment:
By construction, I would say we will never have a duplicate row for a given blockname + pnn. I am not an SQL expert though and I could be wrong. But if I am right, this would make this query faster and cleaner.

Reply:
No. Removing the `group by` statement will make that  query dangerous!